library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/api/RegistryEntryAttribute.java

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.registry.attribute.api;
+
+import com.mojang.serialization.Codec;
+import net.minecraft.resource.ResourceType;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import org.quiltmc.qsl.registry.attribute.impl.RegistryEntryAttributeHolder;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Represents an attribute that is attached to a registry entry.
+ *
+ * @param <R> type of the entries in the registry
+ * @param <V> attached value type
+ */
+@SuppressWarnings("ClassCanBeRecord")
```

Comment:
IIRC, this class was originally supposed to be used in an identity-based map, so you couldn't just create 2 attributes with the same key and have them be "merged". Now that a standard map is used, it can be a record (well, the implementation class can be, anyways).